[9/N][Emit final] Emit final for session window aggregations

guozhangwang opened the pull request:
1. Add a new API for session windows to range query session window by end time (KIP related).
2. Augment session window aggregator with emit strategy.
3. Minor: consolidated some dup classes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

guozhangwang committed:
first commit to have logic ready

guozhangwang committed:
remove dup classes

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java:
```diff
@@ -39,6 +39,13 @@
 public interface SessionStore<K, AGG> extends StateStore, ReadOnlySessionStore<K, AGG> {
 
+    // TODO: javadoc; both ends are inclusive
+    default KeyValueIterator<Windowed<K>, AGG> findSessions(final Instant earliestSessionEndTime,
```
This is related to 1) in the description, and the first open question: is this public API worth to add? Note I added it into SessionStore not ReadOnlySessionStore, to not expose via IQv1, also I've only added this function for `Instant` param type as well.

mjsax replied:
I think there is no way around it? In the end, we allow users to plugin a custom session-store -- thus, if the use the new emit-final, why will need to implement this new method -- existing code with custom session-stores should not break, because existing code does neither implement but also not sure this new method.

If we don't make it public API, we would prevent users to pass in custom session-stores in combination with the new emit-final feature, what seems to be too restrictive?

mjsax replied:
Why did you pick `Instant` over `long` (wondering if `long` might be better as it's more an internal API)?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java:
```diff
@@ -202,25 +205,43 @@ public void remove(final Windowed<Bytes> sessionKey) {
     public byte[] fetchSession(final Bytes key,
-                               final long earliestSessionEndTime,
-                               final long latestSessionStartTime) {
+                               final long sessionStartTime,
```
This is a minor fix on the param names: the old ones are simply wrong and misleading.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java:
```diff
@@ -61,6 +65,18 @@ public <R> QueryResult<R> query(final Query<R> query,
+        final long latestEndTime = ApiUtils.validateMillisecondInstant(latestSessionEndTime,
+                prepareMillisCheckFailMsgPrefix(latestSessionEndTime, "latestSessionEndTime"));
+
+        final KeyValueIterator<Bytes, byte[]> bytesIterator = wrapped().fetchAll(earliestEndTime, latestEndTime);
```
This is the second open question: with the current prefixed (base, i.e. time-first) session key schema, this fetchAll would be effectively searching for `[earliestEnd, INF]` because of this logic: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java#L46

This is because we translate the range query without key inside `AbstractRocksDBTimeOrderedSegmentedBytesStore` by using the `lower/upperRange` instead of `lower/upperRangeFixedSize`): https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBTimeOrderedSegmentedBytesStore.java#L241-L242

I cannot remember why we need to do this. @lihaosky @mjsax do you remember why?

mjsax replied:
Not sure -- I always need to think very hard to understand (not even sure if I succeed) the fetch logic and how we compute the bounds...

But same question as above: why do we need this new method instead of calling `findSessions(null, null, A, B)` -- I briefly dug into the code and it seems it would do the same thing?

guozhangwang replied:
The main reason is that for emit final I need a range API that based on `endTime` for both ends. And that's also why within its implementation I'd have to use `fetchAll` instead of `fetch` here.

mjsax replied:
If I read the code correctly, what `fetchAll()` does is correct: from my understanding, that fetchAll() is implement to find "overlapping sessions" given a lower and upper bound -- the lower bound must be smaller and session end and the upper bound must be smaller than session start to find an overlap. Because the upper bound compares to session start,, and we use the "base" we need to search the full "data/base part" of the store.

I guess the issue is, that you actually cannot use `fetchAll()` at all for our purpose here? Passing in `lastEndTime` does not work (does it) as it would be used to compare to session start-times, but we want to do a comparison to session end time. -- Thus, I think the right solution is, to actually also add the new `findSessions()` to the internal `SegmentedStore` and implement a proper iterator there?

guozhangwang replied:
Sounds good. I would just have a special handling on both lower/upper bound as well as the hasNext function for this specific purpose.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java:
```diff
@@ -91,8 +91,8 @@ public interface SegmentedBytesStore extends StateStore {
      *
-     * @param from the beginning of the time slot from which to search
-     * @param to   the end of the time slot from which to search
+     * @param from the beginning of the time slot from which to search (inclusive)
```
Minor javadoc improvement to remind developers.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimestampedTupleForwarder.java:
```diff
@@ -38,7 +39,7 @@
     TimestampedTupleForwarder(final StateStore store,
                               final ProcessorContext<K, Change<V>> context,
-                              final TimestampedCacheFlushListener<K, V> flushListener,
+                              final CacheFlushListener<K, ?> flushListener,
```
This is 3) in the description: as we can use the base class, we then would not need the duplicated TupleFowarder any more.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,184 @@ public void process(final Record<KIn, VIn> record) {
+            tupleForwarder.maybeForward(new Record<>(windowedkey, new Change<>(newAgg, sendOldValues ? oldAgg : null), newTimestamp));
+        }
+
+        // TODO: consolidate SessionWindow with TimeWindow to merge common functions
```
I realize that our SessionWindow and TimeWindow, and even SlidingWindow caused many code duplications (e.g. here) where we can just consolidate into the same class, with boolean flags indicating if the start/end are inclusive or exclusive, with that we can further reduce code duplication. Will file a JIRA for it.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,184 @@ public void process(final Record<KIn, VIn> record) {
                 store.put(sessionKey, agg);
+
+                maybeForwardUpdate(sessionKey, null, agg, record.timestamp());
+                /*
```
Will remove commented out code when removing WIP, ditto elsewhere.

guozhangwang reviewed (commented):
@lihaosky @mjsax This PR is not mergable yet since I've not committed any new tests. But I'd like to solicit your feedbacks earlier, especially on two open question I made above (one about the public API, one about the inefficiency due to conservative upper/lowerRange with null key).

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K13785-session-aggregation-impl

guozhangwang committed:
do not enable caching with time ordered session

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -83,50 +94,64 @@ private class KStreamSessionWindowAggregateProcessor extends
         private long observedStreamTime = ConsumerRecord.NO_TIMESTAMP;
+        private InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext;
+
+        private final Time time = Time.SYSTEM;
```
Should we not better pass in a `Time` object, so we can mock it using TTD?

guozhangwang replied:
Makes sense. I will do this in a follow-up PR after merging this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,184 @@ public void process(final Record<KIn, VIn> record) {
+            }
+
+            // Update the sent record timestamp to the window end time if possible
+            final long newTimestamp = windowedkey.key() != null ? windowedkey.window().end() : oldTimestamp;
```
For what case could `windowedkey.key() == null` ? Is this even possible?

guozhangwang replied:
This behavior was meant to inherit from the deleted code: https://github.com/apache/kafka/pull/12204/files#diff-85c8c92d464af8eb3a60684bf929725f8fc5263353c38cacc20bee4cefe4fd9eL53, but after checking that logic I now realized it's not necessary anymore (the original PR has to do so since we cannot programmatically guarantee it's always not `null` but in this change we do not have that concern anymore).

Will remove.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java:
```diff
@@ -73,11 +73,10 @@ public void init(final ProcessorContext<Windowed<KIn>, Change<VAgg>> context) {
-        emitFinalLatencySensor = emitFinalLatencySensor(threadId, context.taskId().toString(),
-            internalProcessorContext.currentNode().name(), metrics);
+        emittedRecordsSensor = emittedRecordsSensor(threadId, context.taskId().toString(), processorName, metrics);
+        emitFinalLatencySensor = emitFinalLatencySensor(threadId, context.taskId().toString(), processorName, metrics);
```
Thanks for all the cleanup -- it's somewhat distracting from the actual changes.

Can we (in the future) extract refactorings/cleanups into individual PRs to simplify reviewing?

guozhangwang replied:
Yes! It's my bad to mingle them together here.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java:
```diff
@@ -202,25 +205,43 @@ public void remove(final Windowed<Bytes> sessionKey) {
+                                   null,
+                                    Long.MAX_VALUE,
+                                    endTimeMap.subMap(earliestEndTime, latestEndTime + 1).entrySet().iterator(),
+                                    true);
```
Not sure if I fully understand why we add this new method instead of calling `findSession(null, null, A, B)` ?

The code to create the iterator is different, but I am also not sure why. Is it semantically actually the same? Calling `findSession(null, null, A, B)` would do:
```
registerNewIterator(null, // same
                    null, // same
                    latestSessionStartTime, // why does your code pass Long.MAX_VALUE,
// but because we use `tailMap` instead of `subMap` below, it seems to do the same thing overall?
                    endTimeMap.tailMap(earliestSessionEndTime, true).entrySet().iterator(),
                    true); // same
```

guozhangwang replied:
Logically, the main reason is that for emit-final, we need a range query where the from/to are both `endTime`, i.e. you can see the parameters are `earliestSessionEndTime` and `latestSessionEndTime`.

Whereas for the existing functions, their semantics are based on `earliestSessionEndTime` but `latestSessionStartTime`. And that's also the reason for using `Long.MAX_VALUE` here.

On the physical implementation, the main difference is not in the in-memory session store, but the rocksDB session store. I will reply there separately.

mjsax replied:
Ok. I read the code of `InMemorySessionStore` in detail and now understand what's going on. This LGTM.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K13785-session-aggregation-impl

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java:
```diff
@@ -202,25 +205,43 @@ public void remove(final Windowed<Bytes> sessionKey) {
+        return registerNewIterator(null,
+                                   null,
+                                    Long.MAX_VALUE,
+                                    endTimeMap.subMap(earliestEndTime, latestEndTime + 1).entrySet().iterator(),
```
Nit: can we call `subMap(earliestEndTime, true, latestEndTime, true)` which is the same thing but more "intuitive" as we always search for _inclusive_ bounds throughout the code (otherwise, this is the only place which has an exclusive upper bound).

guozhangwang replied:
Ack.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java:
```diff
@@ -202,25 +205,43 @@ public void remove(final Windowed<Bytes> sessionKey) {
+        // since subMap is exclusive on toKey, we need to plus one
+        return registerNewIterator(null,
+                                   null,
+                                    Long.MAX_VALUE,
```
nit: indention

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K13785-session-aggregation-impl

guozhangwang committed:
github comments

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,182 @@ public void process(final Record<KIn, VIn> record) {
+        }
+
+        private long emitRangeLowerBound() {
+            return lastEmitWindowCloseTime == ConsumerRecord.NO_TIMESTAMP ? 0L : Math.max(0L, lastEmitWindowCloseTime);
```
nit: can this be simplified to `Math.max(0L, lastEmitWindowCloseTime)`? (Can also be address in follow up PR)

guozhangwang replied:
Ack.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,182 @@ public void process(final Record<KIn, VIn> record) {
+                                  final long windowCloseTime,
+                                  final long emitRangeLowerBound,
+                                  final long emitRangeUpperBound) {
+            final long startMs = time.milliseconds();
```
Should we use milliseconds or nanoseconds (I am always unsure)

guozhangwang replied:
Should be ms to be consistent with other metrics.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java:
```diff
@@ -204,7 +203,8 @@ public static void writeBinary(final ByteBuffer buf,
             buf.putLong(endTime);
             buf.putLong(startTime);
-            buf.put(key.get());
+            if (key != null)
```
Can `key` ever be `null` here? (nit: add `{}` to block)

guozhangwang replied:
Yes it's possible since it's used to write lower/upper boundaries in which keys could be null.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java:
```diff
@@ -71,6 +72,36 @@ public byte[] fetchSession(final Bytes key,
+                        final Windowed<Bytes> windowedKey = TimeFirstSessionKeySchema.from(bytes);
+                        final long endTime = windowedKey.window().end();
+
+                        if (endTime <= latestSessionEndTime && endTime >= earliestSessionEndTime)
```
nit: add `{}` to block

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java:
```diff
@@ -35,7 +35,7 @@ public class SessionKeySchema implements SegmentedBytesStore.KeySchema {
 
     public static int keyByteLength(final Bytes key) {
-        return key.get().length + 2 * TIMESTAMP_SIZE;
+        return (key == null ? 0 : key.get().length) + 2 * TIMESTAMP_SIZE;
```
Can `key` ever be `null` here?

guozhangwang replied:
Yes that's possible -- see above comment, for the lower/upper bound cases.

mjsax reviewed (approved):
Overall LGTM.

A few nits. Also, there is some missing JavaDoc and stuff you did just put into comments but not remove yet. Also ok to cleanup in a follow up PR.

guozhangwang commented:
Thanks @mjsax , I've addressed your comment, and also added the test coverage. While adding tests I noticed a bug in the code and fixed it (see my comment above).

Could you please take another look?

guozhangwang committed:
add unit tests, github comments

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:
```diff
@@ -148,55 +173,174 @@ public void process(final Record<KIn, VIn> record) {
+            return windowCloseTime - 1;
+        }
+
+        private boolean shouldRangeFetch(final long emitRangeLowerBound, final long emitRangeUpperBound) {
```
This is one minor bug I detected in the latest commit.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java:
```diff
@@ -71,6 +72,36 @@ public byte[] fetchSession(final Bytes key,
+                                                         final long latestSessionEndTime) {
+        final List<KeyValueSegment> searchSpace = segments.segments(earliestSessionEndTime, latestSessionEndTime, true);
+
+        // here we want [0, latestSE, FF] as the upper bound to cover any possible keys,
```
This is the other minor bug I detected in the latest commit.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java:
```diff
@@ -52,6 +53,16 @@ private SessionStore<Bytes, byte[]> maybeWrapCaching(final SessionStore<Bytes, b
             return inner;
         }
+
+        if (!inner.persistent()) {
```
Here's the other change I made in order to work around the current tricky situation, since in-memory stores are always "time ordered" as well, we stripe the caching if the inner store is not persistent. cc @mjsax

mjsax replied:
If we strip the caching, this applies to eager emitting, right? -- So it would be a behavioral change? Do we want to piggy-back such a change into this KIP? Sounds "risky"?

guozhangwang replied:
Sounds good, I will move this logic into the earlier stage during the topology building phase for the moment. Also cc @lihaosky who would do similar things for sliding windows.

guozhangwang committed:
checkystle and findbugs

guozhangwang committed:
unit tests

mjsax commented on streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java:
```diff
@@ -253,20 +300,29 @@ public void shouldRemoveMergedSessionsFromStateStore() {
 
     @Test
     public void shouldHandleMultipleSessionsAndMerging() {
+        time.sleep(1001L);
```
Might it be easier to change the internal config of the "emit interval" and set it to zero (instead of advancing time)?

guozhangwang replied:
Ack!

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K13785-session-aggregation-impl

guozhangwang committed:
github comments

guozhangwang committed:
github comments 2

mjsax commented:
LGTM (assuming Jenkins passes).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java:
```diff
@@ -286,7 +286,8 @@ private <VR> StoreBuilder<SessionStore<K, VR>> materialize(final MaterializedInt
         }
 
-        if (materialized.cachingEnabled()) {
+        // do not enable cache if the emit final strategy is used
```
cc @lihaosky this is what I did for session store during the topology building phase.
